File: zig/src/runtime/minrt.c
#include <windows.h>

int _fltused = 1;

extern int main(void);

__declspec(dllexport) void print_i64(long long val) {
    char buf[32];
    int len = 0;
    long long tmp = val;
    if (tmp < 0) {
        buf[len++] = '-';
        tmp = -tmp;
    }
    char digits[20];
    int i = 0;
    if (tmp == 0) {
        digits[i++] = '0';
    } else {
        while (tmp > 0) {
            digits[i++] = (char)(tmp % 10) + '0';
            tmp /= 10;
        }
    }
    while (i > 0) {
        buf[len++] = digits[--i];
    }
    buf[len++] = '\n';

    HANDLE handle = GetStdHandle(STD_OUTPUT_HANDLE);
    if (handle == INVALID_HANDLE_VALUE || handle == NULL) return;
    DWORD written;
    WriteFile(handle, buf, len, &written, NULL);
}

__declspec(dllexport) void print_f64(double val) {
    char buf[64];
    int len = 0;
    if (val < 0) {
        buf[len++] = '-';
        val = -val;
    }
    long long whole = (long long)val;
    double frac = val - (double)whole;
    char digits[20];
    int i = 0;
    if (whole == 0) {
        digits[i++] = '0';
    } else {
        long long tmp = whole;
        while (tmp > 0) {
            digits[i++] = (char)(tmp % 10) + '0';
            tmp /= 10;
        }
    }
    while (i > 0) {
        buf[len++] = digits[--i];
    }
    if (frac > 0.000001) {
        buf[len++] = '.';
        for (int d = 0; d < 6 && frac > 0.000001; d++) {
            frac *= 10.0;
            int digit = (int)frac;
            buf[len++] = (char)digit + '0';
            frac -= (double)digit;
        }
    }
    buf[len++] = '\n';
    HANDLE handle = GetStdHandle(STD_OUTPUT_HANDLE);
    if (handle == INVALID_HANDLE_VALUE || handle == NULL) return;
    DWORD written;
    WriteFile(handle, buf, len, &written, NULL);
}

__declspec(dllexport) void print_str(long long ptr) {
    if (ptr == 0) return;
    const char *s = (const char *)(size_t)ptr;
    int len = 0;
    while (s[len]) len++;
    if (len == 0) return;

    HANDLE handle = GetStdHandle(STD_OUTPUT_HANDLE);
    if (handle == INVALID_HANDLE_VALUE || handle == NULL) return;
    DWORD written;
    WriteFile(handle, s, len, &written, NULL);
    WriteFile(handle, "\n", 1, &written, NULL);
}

__declspec(dllexport) void bplus_exit(long long code) {
    ExitProcess((UINT)code);
}

static int launched_by_bpc(void) {
    char buf[2];
    DWORD n = GetEnvironmentVariableA("BPC_RUN", buf, 2);
    return n > 0;
}

static void bplus_pause(void) {
    HANDLE in = GetStdHandle(STD_INPUT_HANDLE);
    if (in == INVALID_HANDLE_VALUE || in == NULL) return;
    DWORD mode;
    if (!GetConsoleMode(in, &mode)) return;

    HANDLE out = GetStdHandle(STD_OUTPUT_HANDLE);
    DWORD written;
    const char msg[] = "\nPress any key to continue...\n";
    if (out != INVALID_HANDLE_VALUE && out != NULL) {
        WriteFile(out, msg, (DWORD)(sizeof(msg) - 1), &written, NULL);
    }
    char ch;
    DWORD read;
    ReadFile(in, &ch, 1, &read, NULL);
}

__declspec(dllexport) void bplus_start(void) {
    int code = (int)main();
    if (!launched_by_bpc()) {
        bplus_pause();
    }
    ExitProcess((UINT)code);
}

File: zig/src/runtime/minrt.zig
const STD_OUTPUT_HANDLE: u32 = 0xFFFFFFF5;

extern fn GetStdHandle(nStdHandle: u32) i64;
extern fn WriteFile(hFile: i64, lpBuffer: [*]const u8, nNumberOfBytesToWrite: u32, lpNumberOfBytesWritten: *u32, lpOverlapped: ?*anyopaque) i32;
extern fn ExitProcess(uExitCode: u32) void;

pub export fn print_i64(val: i64) void {
    @setRuntimeSafety(false);
    var buf: [32]u8 = undefined;
    var len: u32 = 0;
    var tmp: i64 = val;
    if (tmp < 0) {
        buf[0] = '-';
        len = 1;
        tmp = -tmp;
    }
    var i: u32 = 0;
    var digits: [20]u8 = undefined;
    if (tmp == 0) {
        digits[0] = '0';
        i = 1;
    } else {
        while (tmp > 0) {
            const d = @as(u8, @intCast(@rem(tmp, 10)));
            digits[i] = d + '0';
            i += 1;
            tmp = @divTrunc(tmp, 10);
        }
    }
    while (i > 0) {
        i -= 1;
        buf[len] = digits[i];
        len += 1;
    }
    buf[len] = '\n';
    len += 1;

    const handle = GetStdHandle(STD_OUTPUT_HANDLE);
    if (handle == -1 or handle == 0) return;
    var written: u32 = 0;
    _ = WriteFile(handle, &buf, len, &written, null);
}

pub export fn print_str(ptr: i64) void {
    @setRuntimeSafety(false);
    if (ptr == 0) return;
    const s: [*:0]u8 = @ptrFromInt(@as(usize, @intCast(ptr)));
    var len: u32 = 0;
    while (s[@as(usize, @intCast(len))] != 0) : (len += 1) {}
    if (len == 0) return;

    const handle = GetStdHandle(STD_OUTPUT_HANDLE);
    if (handle == -1 or handle == 0) return;
    var written: u32 = 0;
    _ = WriteFile(handle, @as([*]const u8, s), len, &written, null);
    const newline = [1]u8{'\n'};
    _ = WriteFile(handle, &newline, 1, &written, null);
}

pub export fn bplus_exit(code: i64) void {
    ExitProcess(@as(u32, @intCast(code)));
}
